docs/tiltfile_concepts.md

```diff
@@ -144,38 +144,95 @@ can be expanded and investigated.
 your app. However, if you need to configure your Kubernetes resources on top of Tilt's
 automatic assembly, you can do so with a call to [`k8s_resource`](api.html#api.k8s_resource).
-Here we'll discuss the most common configuration arguments: `new_name` and `port_forwards`.
-For discussion of other available arguments, see the [api spec](api.html#api.k8s_resource).
 
-`new_name` allows you to specify a new resource name, in case you do not like the automatically generated one.
+Here we'll discuss the most common configuration options for Kubernetes resources:
+`new_name`, `port_forwards`, and `objects`. For discussion of other available arguments, see
+the [API spec](api.html#api.k8s_resource).
+
+`new_name` allows you to specify a new resource name, in case you do not like the
+automatically generated one:
 
 ```python
 # rename the resource "redis:deployment" to "redis"
-k8s_resource('redis:deployment', new_name='redis')
+k8s_resource(workload='redis:deployment', new_name='redis')
 ```
 
+(Use this pattern to rename individual resources; to programmatically rename all resources,
+see [`workload_to_resource_function`](api.html#api.k8s_resource).)
+
 Tilt also supports a few ways to specify `port_forwards`:
 
 ```python
 # connect localhost:9000 to the default container port
-k8s_resource('frontend', port_forwards=9000)
+k8s_resource(
+  workload='frontend',
+  port_forwards=9000
+)
 
 # connect localhost:9000 to container port 8000
-k8s_resource('frontend', port_forwards='9000:8000')
+k8s_resource(
+  workload='frontend',
+  port_forwards='9000:8000'
+)
 
 # connect localhost:9000 to container port 8000
 # and localhost:9001 to container port 8001
-k8s_resource('frontend', port_forwards=['9000:8000', '9001:8001'])
+k8s_resource(
+  workload='frontend',
+  port_forwards=['9000:8000', '9001:8001']
+)
+```
+
+Additionally, you may want to add additional Kubernetes objects to an existing
+resource (say, group a Secret with the Deployment that makes use of it), or group
+non-workload objects into their own resource (e.g. make a CRD its own resource X,
+so a workload containing an instance of that CR may name X as a dependency). To
+accomplish this, use the `k8s_resource.objects` parameter to specify one or more
+Kubernetes objects.
+
+```python
+# associate an existing Secret and Volume with the "frontend" service
+k8s_resource(
+  workload='frontend',
+  objects=['frontend:secret', 'frontend:volume']
+)
+
+# make a new resource consisting of some objects necessary for cluster setup
+k8s_resource(
+  objects=['my-ns:namespace', 'kafka:crd', 'some-ingress:ingress'],
+  new_name='cluster-setup',
+)
 ```
 
+#### Kubernetes Object Selectors
+
+Tilt specifies Kubernetes objects via _Kubernetes object selectors_.
+The most specific selector for any object is its _fullname_: a colon-separated
+string specifying `$NAME:$KIND:$NAMESPACE` (e.g.: `redis:deployment:default`).
```

Comment:
I think there's also group. There seems to be some confusion about if every object has a group. Apparently in some cases it won't be filled in in the structs ( https://github.com/tilt-dev/tilt/blob/dmiller/ch6889/implement-objects-parameter/internal/k8s/client.go#L417-L434 ), but I think semantically every object has a group.

Reply:
Yeah I think once we figure out what that parameter should be I'll change the docs to add that. It shouldn't affect any other parts of the docs.

Reply:
FWIW, I printed out all of the Groups, Versions and Kinds that we parse out of the YAML used in Tiltfile tests:

```
 [I] ➜ go test -timeout 30s github.com/windmilleng/tilt/internal/tiltfile -run Test -v 2>&1 | grep "Group:"
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: , Version: v1, Kind: Service
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: apps, Version: v1, Kind: Deployment
Group: fission.io, Version: v1, Kind: Environment
Group: fission.io, Version: v1, Kind: Environment
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Secret
Group: , Version: v1, Kind: Namespace
Group: , Version: v1, Kind: Namespace
```

Looks like secrets and namespaces don't have groups which are two specific use cases we know people would want to use the objects parameter.

I think the Kubernetes way to solve this is to use a GroupVersion instead of a group. That would look like this:

```markdown
Tilt specifies Kubernetes objects via _Kubernetes object selectors_.
The most specific selector for any object is its _fullname_: a colon-separated
string specifying `$NAME:$KIND:$NAMESPACE:$GROUPVERSION` (e.g.: `redis:deployment:default:apps/v1` or `foo:secret:default:v1`).
```

WDYT @dbentley?

Reply:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core shows that secret still has a Group.

Reply:
Whoops, I was testing the k8s object matches that we create, not the objects that they were matching against. Here's the correct list:

```
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:52 Group: , Version: v1, GroupVersion: v1, Kind: Service
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:52 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: apps, Version: v1, GroupVersion: apps/v1, Kind: Deployment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: fission.io, Version: v1, GroupVersion: fission.io/v1, Kind: Environment
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Secret
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
2020/05/18 15:10:53 Group: , Version: v1, GroupVersion: v1, Kind: Namespace
```

So I acknowledge that the docs say that the secret has a group, yet the code says that it isn't set. And we have that comment from past me/Matt saying "An empty Group is valid, though: it's empty for apps in the core group." which matches what I'm seeing above. Secrets, namespaces and services don't have a group at runtime, but the docs say that the group should be "core".

Maybe it would be sufficient to say that if a user has specified the group "core", it will match against any resource that has an empty group?

Reply:
Maybe this has something to do with how we ask runtime for the ObjectKind, and then ask that thing for a GroupVersionKind??

Reply:
Only docs I can find on this behavior is [here in the last bullet point](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#review-your-request-attributes)

Reply:
I talked more with Dan in a DM and we decided to leave group off for now and wait until we encounter a use case for specifying group in the wild.

Reply:
Cool, i shall take no action :)